### ISSUE TITLE
Memory usage decrease iniciative

### ISSUE THREAD

**Issue by alxchk** [n1nj4sec/pupy#348]
Currently Linux client eats about 25 MB ram from statring point. After long and resultless measurments and investigations, I came to conclusion that there is simply to many stuff loaded to memory. Most of that is functional objects (functions). I propose to adopt code and dependencies to remove duplicate libraries. These mostly about crypto libraries and parsers (XML/YAML).

Currently pupy code requires only pycrypto to run. In unstable branch I switched pycrypto to pycryptodome. I think at this point we can try to remove all libraries which implements different things missing from pycrypto. Second issue is Hazmat's cryptography. It's quite fat binding over OpenSSL. Are we really require it? Maybe it's possible to migrate such code to pycryptodome? Or maybe it's better to move in opposite direction - from pycryptodome to cryptography?

Second thing are transports. Some of them depends on libraries which are not used in other places (like yaml). I think it's better to adopt them to reduce dependencies.

**Comment by n1nj4sec** [comment]
25Mo of RAM is not that bad is it ? :) 

pupy can run without pycrypto (there is a lightweight, pure python replacement in lib/transports/cryptoutils/pyaes/) that can be used if pycrypto is missing for all transports using AES (but it's slower).
One of the main problem is that a lot of modules uses different crypto libraries (lazagne, scapy, impacket, etc ...).

But I agree we should use only one crypto library at least for pupy internals. I don't know pycryptodome but it seems they only make bugfix releases and cryptography is more famous so I think we should go for cryptography but perhaps you have arguments to go the other way ?
NB: scapy switched from pycrypto to cryptography a few month ago https://github.com/secdev/scapy/pull/371

For yaml I agree it's totally useless (It's here only because scramblesuit (from tor obfsproxy project) uses it)
Same for OpenSSL, pupy network transports doesn't need it it's only some external modules :/
In general,  we should try to avoid using external libraries as much as possible when it's easily feasible using ctypes with reasonable performances ( like in memorpy to answer your python-ptrace question ;))

Perhaps a solution would be to put a mechanism in place to unload some python modules after pupy's module finished ? idk if cleaning sys.modules[module] and call gc.collect() would be enought to free some RAM

**Comment by alxchk** [comment]
> 25Mo of RAM is not that bad is it ? :)

Well, for 1-week shut-and-run projects  it doesn't matter, but for long term pentests reliability and stealth are vital. And hanged "bash" which has size of 80Mb isn't something stealth :)

> pupy can run without pycrypto (there is a lightweight, pure python replacement in lib/transports/cryptoutils/pyaes/) that can be used if pycrypto is missing for all transports using AES (but it's slower).

I think pure python code can stay for some limited cases, but for regular operation it's better to have something native. It's still a question, are native modules really faster in case of usage in python (where still tonns of overhead around), but speed is vital for cases when you need to transfer a lot of data, which are quite regular.

> But I agree we should use only one crypto library at least for pupy internals. I don't know pycryptodome but it seems they only make bugfix releases and cryptography is more famous so I think we should go for cryptography but perhaps you have arguments to go the other way ?

Between cryptography and pycryptodome I'd like to stay with pycryptodome. In normal world where people write software using their brain and feel of beauty cryptography should be better, because it uses openssl as backend (and _ssl and other stuff depends on openssl anyway). But somehow they managed to make _bindings_ to openssl much more larger than openssl library itself. More then that, it's memory usage is horrible.  To compare:

Standard python 2.7.13/Linux + loaded library

|                 | VSS      | RSS   |
|---------------|---------|-------|
|Cryptography| 133852|29040 |
|PyCryptodome| 61780|9308   |

It's amazing. I can't say this fact can surprise me. After observing that shit which is named python standard library not so much things can surprise me. But, anyway, some libs are using cryptography, and there is strong tendency. So, there are three ways to deal with this. 

1. Move pupy to cryptography. It will eat about 40-50 MBs of RAM from the beginning, but futher loading will not increase it.
2. Stay with pycryptodome, and close eyes on double loading. Anyway, not so much modules depends on cryptography for now, and those which are quite fat anyway.
3. Rewrite or make adoptation for modules and libs which depends on cryptography to use pycryptodome (if it's possible of course).

Personally I'd like to stay with (3) or (2).

> Perhaps a solution would be to put a mechanism in place to unload some python modules after pupy's module finished ? idk if cleaning sys.modules[module] and call gc.collect() would be enought to free some RAM

Well, I killed a lot of time to make situation better. You can check in unstable branch (but most things are already merged). Python is very fat by itself, and any loaded resource, any docstring make situation worse.

**Comment by n1nj4sec** [comment]
Yup I saw what you did but you only freed pupyimporter.modules, I just asked if you could imagine a way to free sys.modules by finding a trick or manipulating python internals 
Example : you create a module bigfile.py containing 
```python
A="a"*30*10**6
```
then if you import bigfile your process is 30Mo larger and you can't free it ! Then you can try to 
```
import bigfile
del sys.modules['bigfile']
del bigfile
with open("bigfile.py",'w') as w:
    w.write("")
oldbigfile=bigfile
reload(bigfile)
del oldbigfile
import gc
gc.collect()
```
Nothing works, we can't free the module it just seems python can't natively "unload" a module :/

**Comment by n1nj4sec** [comment]
Ok, I found a way, here is the idea :)
```
import bigfile
from ctypes import *

_PyModule_Clear=pythonapi._PyModule_Clear
_PyModule_Clear.restype = None
_PyModule_Clear.argtypes = [ py_object ]
_PyModule_Clear(bigfile)
```
this actually free the memory and reduces the process RAM consumption

**Comment by alxchk** [comment]
.. And this is a wrong way to go. If you'll take a look into python sources, you'll find that _PyModule_Clear is used in Module object destructor (https://github.com/python/cpython/blob/2.7/Objects/moduleobject.c#L226).

So, it's all about references.

```
import psutil
import os, gc, sys
p = psutil.Process()

with open('lol.py', 'w') as out:
    out.write('A="a"*30*10**6'+'\n')

print "0. RSS: ", p.memory_info().rss
import lol
lol_id = id(sys.modules['lol'])
print 'Module object ID: ', lol_id
print "1. RSS: ", p.memory_info().rss

# We can just delete contents, if we don't know the source of
# reference leakage to get immediate result.
# for k in sys.modules['lol'].__dict__.keys():
#     del sys.modules['lol'].__dict__[k]

gc.collect()

print "2. RSS: ", p.memory_info().rss
del sys.modules['lol']
gc.collect()

print "3. RSS: ", p.memory_info().rss

for obj in gc.get_objects():
    if id(obj) == lol_id:
        print 'Referrers: ', [
            id(x) for x in gc.get_referrers(obj)
        ]

del lol

gc.collect()

print "5. RSS: ", p.memory_info().rss

for obj in gc.get_objects():
    if id(obj) == lol_id:
        print 'Referrers: ', [
            id(x) for x in gc.get_referrers(obj)
        ]
```

```
0. RSS:  12697600
Module object ID:  140243407739656
1. RSS:  42766336
2. RSS:  42766336
3. RSS:  42766336
Referrers:  [140243377852856, 140243409019240]
5. RSS:  12910592
```

**Comment by alxchk** [comment]
BTW, reloading module with lower memory consuption is working in pupy even now. You can easily check this.

```
cat >>packages/all/lol.py
A="a"*30*10**6

run load_package -f lol
run ps -I MEMORY_INFO -s PUPY_PID_HERE

cat >>packages/all/lol.py
A="lol"

run load_package -f lol
run ps -I MEMORY_INFO -s PUPY_PID_HERE
```

**Comment by alxchk** [comment]
I added experemental support for 1-time loading (run -1 lol).

https://github.com/alxchk/pupy/commit/1ebe820b5710f4acb9e21774ae4c2af235df10f6

It's still quite useless, but shows several things which should be taken into account once dealing with RPyC

**Comment by CG-root** [comment]
I have noticed that the Windows payload of Pupy consumes at least 300mb RAM.
Is that normal?

**Comment by CG-root** [comment]
@alxchk , I see that it's consuming a lot of memory like ~300Mb.
Can I choose which modules to load?

**Comment by CG-root** [comment]
@alxchk , is there a way to debug this behavior why it's consuming so much RAM ?

**Comment by alxchk** [comment]
There is no easy way to debug this. You can try to monitor which sequence of actions leads to this. Initial consumption should be around 22-26 mb depends on situation. Then try to figure out which actions trigger memory growing

**Comment by CG-root** [comment]
I think I will skip on that. is it easier now to build your version?
I mean that two months ago I tried and you told me that I need to make a
lot of things in order for that to work

**Comment by alxchk** [comment]
You can try prebuilt docker image. Also
https://github.com/alxchk/pupy/wiki/Installation-(Debian-9)

**Comment by CG-root** [comment]
thanks. I'll try.
on another matter,  I'm trying to bypass some behavior analysis product and
once I open a shell, it says that the process is malicious. is there any
solution except migrating to explorer. exe to resolve that?

**Comment by CG-root** [comment]
@alxchk , I have used the prebuilt option and installed it on Debian 9.
The problem is that the ps1 one liner option still consumes high amount of RAM.
On Windows 10, the powershell.exe process consumes ~200Kb RAM.
Is that normal?

**Comment by alxchk** [comment]
Okay, now its clear.
Im not sure nothing can improve the situation, but yes, this is normal for powershell method. Powershell host with reflected loader eats 300-400 mb of ram. As workaround you can migrate after loading to some other process. 
In case I have spare time I will look what can cause this, but personally Im very sceptical about using powershell as the stage.

**Comment by CG-root** [comment]
I agree. I have now used the exe option and the process consumes ~20Mb only
